File: 3645_maximum_total_from_optimal_activation_order/Solution.cs
// LeetCode 3645 - Maximum Total from Optimal Activation Order
// https://leetcode.com/problems/maximum-total-from-optimal-activation-order/

using System;
using System.Collections.Generic;

public class Solution {
    public long MaxTotal(int[] value, int[] limit) {
        var g = new Dictionary<int, List<int>>();
        for (int i = 0; i < value.Length; i++) {
            if (!g.ContainsKey(limit[i])) g[limit[i]] = new List<int>();
            g[limit[i]].Add(value[i]);
        }
        long ans = 0;
        foreach (var kv in g) {
            int lim = kv.Key;
            var vs = kv.Value;
            vs.Sort((a, b) => b.CompareTo(a));
            for (int i = 0; i < Math.Min(lim, vs.Count); i++) ans += vs[i];
        }
        return ans;
    }
}

File: 3645_maximum_total_from_optimal_activation_order/solution.ts
// LeetCode 3645 - Maximum Total from Optimal Activation Order
// https://leetcode.com/problems/maximum-total-from-optimal-activation-order/

export function maxTotal(value: any, limit: any): any {
    const g = new Map();
    for (let i = 0; i < value.length; i++) {
        if (!g.has(limit[i])) g.set(limit[i], []);
        g.get(limit[i]).push(value[i]);
    }
    let ans = 0;
    for (const [lim, vs] of g) {
        vs.sort((a, b) => b - a);
        for (let i = 0; i < Math.min(lim, vs.length); i++) ans += vs[i];
    }
    return ans;
}
